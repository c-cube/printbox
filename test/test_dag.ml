
let b =
  let uid = ref 0 in
  let id() = incr uid; "["^Int.to_string !uid^"]" in
  let (!) v = `Subtree_with_ID (id(), `Text v) in
  let (+) x y = `Subtree_with_ID (id(), `Tree (`Text "+", [x; y])) in
  let ( * ) x y = `Subtree_with_ID (id(), `Tree (`Text "*", [x; y])) in
  let (/) x y = `Subtree_with_ID (id(), `Tree (`Text "/", [x; y])) in
  let (~-) x = `Subtree_with_ID (id(), `Tree (`Text "-", [x])) in
  let (-) x y = `Subtree_with_ID (id(), `Tree (`Text "-", [x; y])) in
  let f x = `Subtree_with_ID (id(), `Tree (`Text "f", [x])) in
  let a = !"a" in
  let b = !"b" in
  let c =  a + b in
  let d = a * b + b * b * b in
  let c = c + c + !"1" in
  let c = c + !"1" + c + ~-a in
  let d = d + d * !"2" + f (b + a) in
  let d = d + !"3" * d + f (b - a) in
  let e = c - d in
  let f = e * e in
  let g = f / !"2" in
  let g = g + !"10" / f in
  PrintBox.Simple.to_box_centered @@ PrintBox.Simple.reformat_dag 9 g

let () = print_endline @@ PrintBox_text.to_string b
