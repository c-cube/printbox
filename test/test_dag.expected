                                               +
────────────────────────────────────────────────────────────────────────────────────────┬───────
                                           /                                            │   /
──────────────────────────────────────────────────────────────────────────────────────┬─┼──┬────
                                         [29]                                         │2│10│[29]
                                         ────                                         │ │  │
                                         *                                            │ │  │
─────────────────────────────────────────────────────────────────────────────────┬────┤ │  │
                                      [28]                                       │[28]│ │  │
                                      ────                                       │    │ │  │
                                      -                                          │    │ │  │
─────────────────────────────┬───────────────────────────────────────────────────┤    │ │  │
            +                │                       +                           │    │ │  │
────────────────────────┬────┼───────────────────────────────────────────┬───────┤    │ │  │
          +             │-   │                   +                       │ f     │    │ │  │
───────────────────┬────┼────┼─────────────────────────────────┬─────────┼───────┤    │ │  │
       [13]        │[10]│[1] │              [21]               │  [25]   │ [22]  │    │ │  │
       ────        │    │    │              ────               │  ────   │ ────  │    │ │  │
       +           │    │    │              +                  │  *      │ -     │    │ │  │
──────────────┬────┤    │    ├─────────────────────────┬───────┼────┬────┼───┬───┤    │ │  │
     [10]     │[12]│    │    │          [20]           │ [17]  │[24]│[21]│[2]│[1]│    │ │  │
     ────     ├────┤    │    │          ────           │ ────  ├────┤    │   │   │    │ │  │
     +        │1   │    │    │          +              │ f     │3   │    │   │   │    │ │  │
──────────┬───┤    │    │    ├────────────────┬────────┼───────┤    │    │   │   │    │ │  │
   [9]    │[8]│    │    │    │      [7]       │  [19]  │ [16]  │    │    │   │   │    │ │  │
   ───    ├───┤    │    │    │      ───       │  ────  │ ────  │    │    │   │   │    │ │  │
   +      │1  │    │    │    │      +         │  *     │ +     │    │    │   │   │    │ │  │
──────┬───┤   │    │    │    ├──────┬─────────┼───┬────┼───┬───┤    │    │   │   │    │ │  │
[3]   │[3]│   │    │    │    │[6]   │[5]      │[7]│[18]│[2]│[1]│    │    │   │   │    │ │  │
───   │   │   │    │    │    ├───   ├───      │   ├────┤   │   │    │    │   │   │    │ │  │
+     │   │   │    │    │    │*     │*        │   │2   │   │   │    │    │   │   │    │ │  │
├─[1] │   │   │    │    │    │├─[1] │├─[4]    │   │    │   │   │    │    │   │   │    │ │  │
│ ─── │   │   │    │    │    │└─[2] ││ ───    │   │    │   │   │    │    │   │   │    │ │  │
│ a   │   │   │    │    │    │      ││ *      │   │    │   │   │    │    │   │   │    │ │  │
└─[2] │   │   │    │    │    │      ││ ├─[2]  │   │    │   │   │    │    │   │   │    │ │  │
  ─── │   │   │    │    │    │      ││ └─[2]  │   │    │   │   │    │    │   │   │    │ │  │
  b   │   │   │    │    │    │      │└─[2]    │   │    │   │   │    │    │   │   │    │ │  │
